### DISCUSSION TITLE
how to generate public token and handle errors.

### DISCUSSION THREAD

**Discussion by Nader-CS** [vercel/next.js#74330] [App Router]
Hello, my app needs to have a public token saved in a cookie for subsequent use. It also needs to attach this public token to client-side queries and handle API request errors. How can I implement this?

**Comment by barnejs** [comment]
store any tokens on client-side is bad practice. Are you sure you dont have any other solutions?

**Reply by Nader-CS** [reply]
This is a public token, not sensitive. My backend requires every request to include a public token. Sometimes, I'm in a client component and need to make a request. What is your opinion?

**Comment by wilsonadenuga** [comment]
You can use cookies, on the server, you can access it as described [here](https://nextjs.org/docs/app/api-reference/functions/cookies)
on the client, you can use a third-party library, e.g [js-cookie](https://www.npmjs.com/package/js-cookie)

**Reply by Nader-CS** [reply]
no you don't understand me , if i generate token in client side , server component runs first and then will request to server without public token , if i don it in server , i can't set cookies in server component

**Comment by Ekpo-Emmanuel** [comment]
1. Create a server-side API route to generate a public token and set it in an HttpOnly cookie.
2. Use a client-side hook that retrieves the token from cookies and provides a method to make authenticated requests. This hook will handle token attachment and basic error management.
3. In your client components, utilize this hook to fetch data with the token automatically included. Implement try/catch blocks to handle potential request failures gracefully.

The key is to generate the token server-side, store it securely in a cookie, and create a reusable method for attaching it to requests. This approach prevents token generation issues in server components and provides a clean, consistent way to handle public token authentication.
